### PR TITLE
PR: gh#128 : Adding a script to compare the functions

### PR DESCRIPTION
**Pull request by kanjoe24**
A script file is being added, which will run as part of makefile that will display information if there are no tests available in test suite for a function present in header file

### CODE REVIEW

**Comment by kanjoe24** [comment]
logs from device settings:

```
Generating lib [/home/jpn323/workspace/gh128-compare-script/scripts/rdk-halif-device_settings-ubuntu/ut/ut-core/framework/ut-control/build/linux/lib/libut_control.so]
make[3]: Leaving directory '/home/jpn323/workspace/gh128-compare-script/scripts/rdk-halif-device_settings-ubuntu/ut/ut-core/framework/ut-control'
make[2]: Leaving directory '/home/jpn323/workspace/gh128-compare-script/scripts/rdk-halif-device_settings-ubuntu/ut/ut-core/framework/ut-control'
Framework downloaded and built
make[2]: Entering directory '/home/jpn323/workspace/gh128-compare-script/scripts/rdk-halif-device_settings-ubuntu/ut/ut-core'
TARGET_EXEC [hal_test]
TARGET [linux]
COMPILER [ gcc -ggdb -o0 -Wall]
VARIANT [C]
Building [/home/jpn323/workspace/gh128-compare-script/scripts/rdk-halif-device_settings-ubuntu/ut/ut-core/framework/CUnit-2.1-3/CUnit/Sources/Framework/TestRun.c]
Building [/home/jpn323/workspace/gh128-compare-script/scripts/rdk-halif-device_settings-ubuntu/ut/ut-core/src/c_source/cunit_lgpl/ut_automated.c]
Building [/home/jpn323/workspace/gh128-compare-script/scripts/rdk-halif-device_settings-ubuntu/ut/ut-core/src/c_source/cunit_lgpl/ut_basic.c]
Building [/home/jpn323/workspace/gh128-compare-script/scripts/rdk-halif-device_settings-ubuntu/ut/ut-core/src/c_source/cunit_lgpl/ut_console.c]
Building [/home/jpn323/workspace/gh128-compare-script/scripts/rdk-halif-device_settings-ubuntu/ut/ut-core/src/c_source/ut_cunit.c]
Building [/home/jpn323/workspace/gh128-compare-script/scripts/rdk-halif-device_settings-ubuntu/ut/ut-core/src/ut_kvp_profile.c]
Building [/home/jpn323/workspace/gh128-compare-script/scripts/rdk-halif-device_settings-ubuntu/ut/ut-core/src/ut_main.c]
Linking test /home/jpn323/workspace/gh128-compare-script/scripts/rdk-halif-device_settings-ubuntu/ut/build/linux/obj/hal_test
make[2]: Leaving directory '/home/jpn323/workspace/gh128-compare-script/scripts/rdk-halif-device_settings-ubuntu/ut/ut-core'
ut-control LIB Copied to [/home/jpn323/workspace/gh128-compare-script/scripts/rdk-halif-device_settings-ubuntu/ut/bin]
*******************************************************************************
  ** Function 'dsGetHdmiVersion' is NOT found in ANY object file! **
*******************************************************************************
*******************************************************************************
  ** Function 'dsGetAllmEnabled' is NOT found in ANY object file! **
*******************************************************************************
*******************************************************************************
  ** Function 'dsSetAllmEnabled' is NOT found in ANY object file! **
*******************************************************************************
make[1]: Leaving directory '/home/jpn323/workspace/gh128-compare-script/scripts/rdk-halif-device_settings-ubuntu/ut/ut-core'
make: Leaving directory '/home/jpn323/workspace/gh128-compare-script/scripts/rdk-halif-device_settings-ubuntu/ut'
jpn323@janus ~/workspace/gh128-compare-script/scripts/rdk-halif-device_settings-ubuntu (main)$ 

```

**Comment by kanjoe24** [comment]
Logs from linux build:
```
develop)$ ./build_ut.sh 
ut-core version selected:[3.2.4]
--- New Version of ut-core released [4.1.0] consider upgrading ---
make: Entering directory '/home/jpn323/workspace/rdk-halif-device_settings/ut'
TARGET NOT SET 
TARGET FORCED TO Linux
TARGET [linux]
UT [build]
make -C ./ut-core TARGET=linux
make[1]: Entering directory '/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core'
TARGET_EXEC [hal_test]
TARGET [linux]
COMPILER [ gcc -ggdb -o0 -Wall]
VARIANT [C]
Ensure ut-core frameworks are present
TARGET= [linux] from [/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core/build.sh]
VARIANT= [C] from [/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core/build.sh]
Framework ut-control already exists
ut-control version selected:[1.6.0]
Third party libraries are built for linux
Completed
Entering ut-control [TARGET=linux]
make[2]: Entering directory '/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core/framework/ut-control'
Building [libut_control.so]
TARGET [linux]
Ensure ut-control framework is present
[/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core/framework/ut-control/configure.sh] TARGET [linux]
Framework [asprintf] already exists
Framework [libfyaml] already exists
~/workspace/rdk-halif-device_settings/ut/ut-core/framework/ut-control ~/workspace/rdk-halif-device_settings/ut/ut-core/framework/ut-control
CMake is installed
openssl.pc, libssl.a, and libcrypto.a are found for linux.
libcurl.a not found
curl.h not found
Framework [libwebsockets] already exists
Framework [libwebsockets] already built for linux
Framework [curl] already exists
Framework [curl] already built for linux
make lib
make[3]: Entering directory '/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core/framework/ut-control'
Building [libut_control.so]
TARGET [linux]
Generating lib [/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core/framework/ut-control/build/linux/lib/libut_control.so]
make[3]: Leaving directory '/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core/framework/ut-control'
make[2]: Leaving directory '/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core/framework/ut-control'
Framework downloaded and built
make[2]: Entering directory '/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core'
TARGET_EXEC [hal_test]
TARGET [linux]
COMPILER [ gcc -ggdb -o0 -Wall]
VARIANT [C]
Linking test /home/jpn323/workspace/rdk-halif-device_settings/ut/build/linux/obj/hal_test
make[2]: Leaving directory '/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core'
ut-control LIB Copied to [/home/jpn323/workspace/rdk-halif-device_settings/ut/bin]
***************************************************************************************
  ** Function 'dsGetHdmiVersion' is NOT found in ANY object file on target linux! **
***************************************************************************************
***************************************************************************************
  ** Function 'dsGetAllmEnabled' is NOT found in ANY object file on target linux! **
***************************************************************************************
***************************************************************************************
  ** Function 'dsSetAllmEnabled' is NOT found in ANY object file on target linux! **
***************************************************************************************
make[1]: Leaving directory '/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core'
```

**Comment by kanjoe24** [comment]
Note : Observed that the build time increases by about 2 mins when built for arm.

File: compare-functions-in-headers-testsuite.sh [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    # Loop through all object files in the object folder
+    for object_file in "$object_folder"/*.o; do
+      # Check if the object file exists
+      if [ ! -f "$object_file" ]; then
+        echo "Warning: No object files found in '$object_folder'."
+        break
+      fi
+
+      # Check if the function is found in the object file (defined or undefined)
+      if nm "$object_file" | grep -w " $func$" > /dev/null || nm "$object_file" | grep -w " U $func$" > /dev/null; then
```
[the remark this answers is not in the document]
Comment (Ulrond) [reply compare-functions-in-headers-testsuite.sh:89]: only check the binary `hal_test`

**Review by Ulrond — CHANGES_REQUESTED** [review]
only check the final binary, no need to check *.o

**Comment by kanjoe24** [comment]
Logs for both linux and arm after changes:

Screenshot 2024-11-18 at 11 23 19
Screenshot 2024-11-18 at 11 27 04

**Comment by Ulrond** [comment]
Move from

```bash
***************************************************************************************
  ** Function 'dsGetHdmiVersion' is NOT found in ANY object file on target linux! **
***************************************************************************************
***************************************************************************************
  ** Function 'dsGetAllmEnabled' is NOT found in ANY object file on target linux! **
***************************************************************************************
***************************************************************************************
  ** Function 'dsSetAllmEnabled' is NOT found in ANY object file on target linux! **
***************************************************************************************
```

to 

```bash
** Warning: Function 'dsGetHdmiVersion()' is NOT found on target linux! **
** Warning: Function 'dsGetAllmEnabled()' is NOT found on target linux! **
** Warning: Function 'dsSetAllmEnabled()' is NOT found on target linux! **
```

**Comment by kanjoe24** [comment]
> Move from
> 
> ```shell
> ***************************************************************************************
>   ** Function 'dsGetHdmiVersion' is NOT found in ANY object file on target linux! **
> ***************************************************************************************
> ***************************************************************************************
>   ** Function 'dsGetAllmEnabled' is NOT found in ANY object file on target linux! **
> ***************************************************************************************
> ***************************************************************************************
>   ** Function 'dsSetAllmEnabled' is NOT found in ANY object file on target linux! **
> ***************************************************************************************
> ```
> 
> to
> 
> ```shell
> ** Warning: Function 'dsGetHdmiVersion()' is NOT found on target linux! **
> ** Warning: Function 'dsGetAllmEnabled()' is NOT found on target linux! **
> ** Warning: Function 'dsSetAllmEnabled()' is NOT found on target linux! **
> ```


Logs with changes:

```shell
cp /home/jpn323/workspace/rdk-halif-device_settings/ut/libs/*.so* /home/jpn323/workspace/rdk-halif-device_settings/ut/bin
make[2]: Leaving directory '/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core'
ut-control LIB Copied to [/home/jpn323/workspace/rdk-halif-device_settings/ut/bin]
  ** Warning Function 'dsGetHdmiVersion' is NOT found on target arm! **
  ** Warning Function 'dsGetAllmEnabled' is NOT found on target arm! **
  ** Warning Function 'dsSetAllmEnabled' is NOT found on target arm! **
make[1]: Leaving directory '/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core'
``` 

```shell
Linking test /home/jpn323/workspace/rdk-halif-device_settings/ut/build/linux/obj/hal_test
make[2]: Leaving directory '/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core'
ut-control LIB Copied to [/home/jpn323/workspace/rdk-halif-device_settings/ut/bin]
  ** Warning Function 'dsGetHdmiVersion' is NOT found on target linux! **
  ** Warning Function 'dsGetAllmEnabled' is NOT found on target linux! **
  ** Warning Function 'dsSetAllmEnabled' is NOT found on target linux! **
make[1]: Leaving directory '/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core'
make: Leaving directory '/home/jpn323/workspace/rdk-halif-device_settings/ut'
```

**Review by Ulrond — APPROVED** [review]
just test it with your release test script also see if it's ok.

**Comment by kanjoe24** [comment]
ut-core release script results

```shell
./release-test-script-ut-core.sh -t feature/gh128-compare-script
UT_CORE_BRANCH_NAME = feature/gh128-compare-script
UT_CONTROL_BRANCH_NAME = 
UT_CONTROL_BRANCH_NAME is empty
==========================================================
RESULTS for ubuntu 
On the branch feature/gh128-compare-script
. PASS
framework/ut-control/build/linux/curl/lib/libcurl.a exists. PASS
Openssl static lib does not exist. PASS 
CMake host binary does not exist. PASS 
build/linux/cpp_libs/lib/libgtest.a exists. PASS
build/linux/cpp_libs/lib/libgtest_main.a exists. PASS
==========================================================
==========================================================
RESULTS for VM-SYNC 
On the branch feature/gh128-compare-script
. PASS
framework/ut-control/build/linux/curl/lib/libcurl.a exists. PASS
framework/ut-control/build/linux/openssl/lib/libssl.a exists. PASS 
framework/ut-control/host-tools/CMake-3.30.0/build/bin/cmake exists. PASS 
build/linux/cpp_libs/lib/libgtest.a exists. PASS
build/linux/cpp_libs/lib/libgtest_main.a exists. PASS
==========================================================
==========================================================
RESULTS for dunfell_arm 
On the branch feature/gh128-compare-script
. PASS
framework/ut-control/build/arm/curl/lib/libcurl.a exists. PASS
framework/ut-control/build/arm/openssl/lib/libssl.a exists. PASS 
CMake host binary does not exist. PASS 
build/arm/cpp_libs/lib/libgtest.a exists. PASS
build/arm/cpp_libs/lib/libgtest_main.a exists. PASS
==========================================================
==========================================================
RESULTS for dunfell_linux 
On the branch feature/gh128-compare-script
. PASS
framework/ut-control/build/linux/curl/lib/libcurl.a exists. PASS
Openssl static lib does not exist. PASS 
CMake host binary does not exist. PASS 
build/linux/cpp_libs/lib/libgtest.a exists. PASS
build/linux/cpp_libs/lib/libgtest_main.a exists. PASS
==========================================================
==========================================================
RESULTS for kirkstone_arm 
On the branch feature/gh128-compare-script
. PASS
framework/ut-control/build/arm/curl/lib/libcurl.a exists. PASS
framework/ut-control/build/arm/openssl/lib/libssl.a exists. PASS 
CMake host binary does not exist. PASS 
build/arm/cpp_libs/lib/libgtest.a exists. PASS
build/arm/cpp_libs/lib/libgtest_main.a exists. PASS
==========================================================
==========================================================
RESULTS for kirkstone_linux 
On the branch feature/gh128-compare-script
. PASS
framework/ut-control/build/linux/curl/lib/libcurl.a exists. PASS
Openssl static lib does not exist. PASS 
CMake host binary does not exist. PASS 
build/linux/cpp_libs/lib/libgtest.a exists. PASS
build/linux/cpp_libs/lib/libgtest_main.a exists. PASS
==========================================================

```

**Comment by kanjoe24** [comment]
platform release script results with ./build_ut.sh TARGET=$TARGET

```shell
./release-test-script-platform.sh -t feature/gh128-compare-script
UT_CORE_BRANCH_NAME = feature/gh128-compare-script
UT_CONTROL_BRANCH_NAME = 
UT_CONTROL_BRANCH_NAME is empty
==========================================================
RESULTS for ubuntu 
On the branch feature/gh128-compare-script
            . PASS
/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core/scripts/rdk-halif-hdmi_cec-ubuntu
Hal binary does not exist. FAIL 
ut/ut-core/framework/ut-control/build/linux/curl/lib/libcurl.a exists. PASS
Openssl static lib does not exist. PASS 
CMake host binary does not exist. PASS 
==========================================================
==========================================================
RESULTS for VM-SYNC 
On the branch feature/gh128-compare-script
            . PASS
/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core/scripts/rdk-halif-hdmi_cec-VM-SYNC
Hal binary does not exist. FAIL 
CURL static lib does not exist. FAIL 
Openssl static lib does not exist. FAIL 
CMake host binary does not exist. FAIL 
==========================================================
==========================================================
RESULTS for dunfell_arm 
On the branch feature/gh128-compare-script
            . PASS
/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core/scripts/rdk-halif-hdmi_cec-dunfell_arm
ut/bin/hal_test exists. PASS
ut/ut-core/framework/ut-control/build/arm/curl/lib/libcurl.a exists. PASS
ut/ut-core/framework/ut-control/build/arm/openssl/lib/libssl.a exists. PASS 
CMake host binary does not exist. PASS 
==========================================================
==========================================================
RESULTS for dunfell_linux 
Branch is not switched. FAIL
/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core/scripts/rdk-halif-hdmi_cec-dunfell_linux
Hal binary does not exist. FAIL 
CURL static lib does not exist. PASS 
Openssl static lib does not exist. PASS 
CMake host binary does not exist. PASS 
==========================================================
==========================================================
RESULTS for kirkstone_arm 
On the branch feature/gh128-compare-script
            . PASS
/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core/scripts/rdk-halif-hdmi_cec-kirkstone_arm
ut/bin/hal_test exists. PASS
ut/ut-core/framework/ut-control/build/arm/curl/lib/libcurl.a exists. PASS
ut/ut-core/framework/ut-control/build/arm/openssl/lib/libssl.a exists. PASS 
CMake host binary does not exist. PASS 
==========================================================
==========================================================
RESULTS for kirkstone_linux 
On the branch feature/gh128-compare-script
            . PASS
/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core/scripts/rdk-halif-hdmi_cec-kirkstone_linux
ut/bin/hal_test exists. PASS
ut/ut-core/framework/ut-control/build/linux/curl/lib/libcurl.a exists. PASS
Openssl static lib does not exist. PASS 
CMake host binary does not exist. PASS 
==========================================================

``` 

platform release script results with ./build_ut.sh for linux and ./build_ut.sh TARGET=arm on arm
```shell
./release-test-script-platform.sh -t feature/gh128-compare-script
UT_CORE_BRANCH_NAME = feature/gh128-compare-script
UT_CONTROL_BRANCH_NAME = 
UT_CONTROL_BRANCH_NAME is empty
==========================================================
RESULTS for ubuntu 
On the branch feature/gh128-compare-script
            . PASS
/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core/scripts/rdk-halif-hdmi_cec-ubuntu
ut/bin/hal_test exists. PASS
ut/ut-core/framework/ut-control/build/linux/curl/lib/libcurl.a exists. PASS
Openssl static lib does not exist. PASS 
CMake host binary does not exist. PASS 
==========================================================
==========================================================
RESULTS for VM-SYNC 
On the branch feature/gh128-compare-script
            . PASS
/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core/scripts/rdk-halif-hdmi_cec-VM-SYNC
ut/bin/hal_test exists. PASS
ut/ut-core/framework/ut-control/build/linux/curl/lib/libcurl.a exists. PASS
ut/ut-core/framework/ut-control/build/linux/openssl/lib/libssl.a exists. PASS 
ut/ut-core/framework/ut-control/host-tools/CMake-3.30.0/build/bin/cmake exists. PASS 
==========================================================
==========================================================
RESULTS for dunfell_arm 
On the branch feature/gh128-compare-script
            . PASS
/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core/scripts/rdk-halif-hdmi_cec-dunfell_arm
ut/bin/hal_test exists. PASS
ut/ut-core/framework/ut-control/build/arm/curl/lib/libcurl.a exists. PASS
ut/ut-core/framework/ut-control/build/arm/openssl/lib/libssl.a exists. PASS 
CMake host binary does not exist. PASS 
==========================================================
==========================================================
RESULTS for dunfell_linux 
On the branch feature/gh128-compare-script
            . PASS
/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core/scripts/rdk-halif-hdmi_cec-dunfell_linux
ut/bin/hal_test exists. PASS
CURL static lib does not exist. PASS 
Openssl static lib does not exist. PASS 
CMake host binary does not exist. PASS 
==========================================================
==========================================================
RESULTS for kirkstone_arm 
On the branch feature/gh128-compare-script
            . PASS
/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core/scripts/rdk-halif-hdmi_cec-kirkstone_arm
ut/bin/hal_test exists. PASS
ut/ut-core/framework/ut-control/build/arm/curl/lib/libcurl.a exists. PASS
ut/ut-core/framework/ut-control/build/arm/openssl/lib/libssl.a exists. PASS 
CMake host binary does not exist. PASS 
==========================================================
==========================================================
RESULTS for kirkstone_linux 
On the branch feature/gh128-compare-script
            . PASS
/home/jpn323/workspace/rdk-halif-device_settings/ut/ut-core/scripts/rdk-halif-hdmi_cec-kirkstone_linux
ut/bin/hal_test exists. PASS
ut/ut-core/framework/ut-control/build/linux/curl/lib/libcurl.a exists. PASS
Openssl static lib does not exist. PASS 
CMake host binary does not exist. PASS 
==========================================================

```

**Review by Ulrond — APPROVED** [review]

**Pull request merged**